# Explore here

### We import the necessary modules for the project.

In [1]:
import requests
import os
import pandas as pd

### We now download and store the data into /data/raw/*

In [2]:
# We specify the file we are downloading and where it is going to be saved.
url = "https://raw.githubusercontent.com/4GeeksAcademy/logistic-regression-project-tutorial/main/bank-marketing-campaign-data.csv"
output_path = "/workspace/ML-Exercise/data/raw/data.csv"

# We will use 
response = requests.get(url)
with open(output_path, 'wb') as my_file:
    my_file.write(response.content)

print(f"File downloaded to {output_path}")

File downloaded to /workspace/ML-Exercise/data/raw/data.csv


### We now generate a dataframe to work with it.

In [3]:
df = pd.read_csv('/workspace/ML-Exercise/data/raw/data.csv', sep =";")
pd.set_option('display.max_columns', None)
df.head()

age        job  marital    education  default housing loan    contact  \
0   56  housemaid  married     basic.4y       no      no   no  telephone   
1   57   services  married  high.school  unknown      no   no  telephone   
2   37   services  married  high.school       no     yes   no  telephone   
3   40     admin.  married     basic.6y       no      no   no  telephone   
4   56   services  married  high.school       no      no  yes  telephone   

  month day_of_week  duration  campaign  pdays  previous     poutcome  \
0   may         mon       261         1    999         0  nonexistent   
1   may         mon       149         1    999         0  nonexistent   
2   may         mon       226         1    999         0  nonexistent   
3   may         mon       151         1    999         0  nonexistent   
4   may         mon       307         1    999         0  nonexistent   

   emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  nr.employed   y  
0           1.1          93.994          -36.4      4.857       5191.0  no  
1           1.1          93.994          -36.4      4.857       5191.0  no  
2           1.1          93.994          -36.4      4.857       5191.0  no  
3           1.1          93.994          -36.4      4.857       5191.0  no  
4           1.1          93.994          -36.4      4.857       5191.0  no

In [4]:
df.tail()

age          job  marital            education default housing loan  \
41183   73      retired  married  professional.course      no     yes   no   
41184   46  blue-collar  married  professional.course      no      no   no   
41185   56      retired  married    university.degree      no     yes   no   
41186   44   technician  married  professional.course      no      no   no   
41187   74      retired  married  professional.course      no     yes   no   

        contact month day_of_week  duration  campaign  pdays  previous  \
41183  cellular   nov         fri       334         1    999         0   
41184  cellular   nov         fri       383         1    999         0   
41185  cellular   nov         fri       189         2    999         0   
41186  cellular   nov         fri       442         1    999         0   
41187  cellular   nov         fri       239         3    999         1   

          poutcome  emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  \
41183  nonexistent          -1.1          94.767          -50.8      1.028   
41184  nonexistent          -1.1          94.767          -50.8      1.028   
41185  nonexistent          -1.1          94.767          -50.8      1.028   
41186  nonexistent          -1.1          94.767          -50.8      1.028   
41187      failure          -1.1          94.767          -50.8      1.028   

       nr.employed    y  
41183       4963.6  yes  
41184       4963.6   no  
41185       4963.6   no  
41186       4963.6  yes  
41187       4963.6   no

In [5]:
# We check the shape to have a better image of the dataframe:
df.shape


(41188, 21)

In [6]:
# We check the information of the dataframe:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41188 entries, 0 to 41187
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             41188 non-null  int64  
 1   job             41188 non-null  object 
 2   marital         41188 non-null  object 
 3   education       41188 non-null  object 
 4   default         41188 non-null  object 
 5   housing         41188 non-null  object 
 6   loan            41188 non-null  object 
 7   contact         41188 non-null  object 
 8   month           41188 non-null  object 
 9   day_of_week     41188 non-null  object 
 10  duration        41188 non-null  int64  
 11  campaign        41188 non-null  int64  
 12  pdays           41188 non-null  int64  
 13  previous        41188 non-null  int64  
 14  poutcome        41188 non-null  object 
 15  emp.var.rate    41188 non-null  float64
 16  cons.price.idx  41188 non-null  float64
 17  cons.conf.idx   41188 non-null 

# From the data we have we will now check the information that might not be relevant at all for the goal we have in our prediction.
    1 - age. Age of customer (numeric) - YES
    2 - job. Type of job (categorical) - YES
    3 - marital. Marital status (categorical) - YES
    4 - education. Level of education (categorical) - YES
    5 - default. Do you currently have credit (categorical) - YES
    6 - housing. Do you have a housing loan (categorical) - YES
    7 - loan. Do you have a personal loan? (categorical) - YES
    8 - contact. Type of contact communication (categorical) - MAYBE 
    9 - month. Last month in which you have been contacted (categorical) - MAYBE
    10 - day_of_week. Last day on which you have been contacted (categorical) - MAYBE    
    11 - duration. Duration of previous contact in seconds (numeric) - MAYBE
    12 - campaign. Number of contacts made during this campaign to the customer (numeric) - MAYBE
    13 - pdays. Number of days that elapsed since the last campaign until the customer was contacted (numeric) - MAYBE
    14 - previous. Number of contacts made during the previous campaign to the customer (numeric) - MAYBE
    15 - poutcome. Result of the previous marketing campaign (categorical) - MAYBE
    16 - emp.var.rate. Employment variation rate. Quarterly indicator (numeric) - YES
    17 - cons.price.idx. Consumer price index. Monthly indicator (numeric) - YES
    18 - cons.conf.idx. Consumer confidence index. Monthly indicator (numeric) - YES
    19 - euribor3m. EURIBOR 3-month rate. Daily indicator (numeric) - MAYBE [WE HAVE TO CHECK THIS ONE TO FIND OUT IT IS A PERSONAL VALUE OR A GENERAL VALUE, IN WHICH CASE WOULD BE A NO.]
    20 - nr.employed. Number of employees. Quarterly indicator (numeric) - MAYBE [NUMBER OF EMPLOYEES OF THE BANK OR THAT HE MANAGES? PROBABLE IT IS A NO]

These decisions were made by looking at the description of the column title and its explanation. Those that have a MAYBE must be examined to determine if they will be relevant or not.

In [7]:
# We will check if there is any column that has a unique variable.
df.nunique()

age                 78
job                 12
marital              4
education            8
default              3
housing              3
loan                 3
contact              2
month               10
day_of_week          5
duration          1544
campaign            42
pdays               27
previous             8
poutcome             3
emp.var.rate        10
cons.price.idx      26
cons.conf.idx       26
euribor3m          316
nr.employed         11
y                    2
dtype: int64

As there are no columns with a unique variable, we will proceed to check the value counts of each column to obtain a clearer view of the data we will work with, as well as identify inconsistencies.

In [8]:
# We generate a loop to iterate on each column.
for column in df.columns:
    print(f"Value counts for {column}:")
    print(df[column].value_counts())
    print("-" * 40)

Value counts for age:
age
31    1947
32    1846
33    1833
36    1780
35    1759
      ... 
91       2
98       2
95       1
87       1
94       1
Name: count, Length: 78, dtype: int64
----------------------------------------
Value counts for job:
job
admin.           10422
blue-collar       9254
technician        6743
services          3969
management        2924
retired           1720
entrepreneur      1456
self-employed     1421
housemaid         1060
unemployed        1014
student            875
unknown            330
Name: count, dtype: int64
----------------------------------------
Value counts for marital:
marital
married     24928
single      11568
divorced     4612
unknown        80
Name: count, dtype: int64
----------------------------------------
Value counts for education:
education
university.degree      12168
high.school             9515
basic.9y                6045
professional.course     5243
basic.4y                4176
basic.6y                2292
unknown             

duration
85      170
90      170
136     168
73      167
124     164
       ... 
1356      1
1103      1
1992      1
3183      1
1399      1
Name: count, Length: 1544, dtype: int64
----------------------------------------
Value counts for campaign:
campaign
1     17642
2     10570
3      5341
4      2651
5      1599
6       979
7       629
8       400
9       283
10      225
11      177
12      125
13       92
14       69
17       58
16       51
15       51
18       33
20       30
19       26
21       24
22       17
23       16
24       15
27       11
29       10
28        8
26        8
25        8
30        7
31        7
35        5
33        4
32        4
34        3
43        2
40        2
42        2
39        1
56        1
37        1
41        1
Name: count, dtype: int64
----------------------------------------
Value counts for pdays:
pdays
999    39673
3        439
6        412
4        118
9         64
2         61
7         60
12        58
10        52
5         46
13        3

The columns month, day_of_week, duration and campaign are results obtained after having contact with the customer during this campaign. Because we are aiming to predict before contacting our customer if he is a potential candidate, this information would not be available for the person aiming to predict.

Therefore we will drop those columns.

In [9]:
df.drop(columns = ["month","day_of_week","duration","campaign"], inplace = True)


The default column has only 3 instances marked as 'Yes' out of 41,188 entries, with the remaining values being 'No' (32,588) and 'Unknown' (8,597). Although the 'Yes' cases are rare, it’s possible that having a default doesn’t necessarily mean a customer will be excluded from selection. It’s likely that much of the database was previously filtered, discarding these customers before any contact attempts were made. We will visually inspect the data for the 'Unknown' and 'Yes' responses to explore this further.

In [10]:
df.head()

age        job  marital    education  default housing loan    contact  \
0   56  housemaid  married     basic.4y       no      no   no  telephone   
1   57   services  married  high.school  unknown      no   no  telephone   
2   37   services  married  high.school       no     yes   no  telephone   
3   40     admin.  married     basic.6y       no      no   no  telephone   
4   56   services  married  high.school       no      no  yes  telephone   

   pdays  previous     poutcome  emp.var.rate  cons.price.idx  cons.conf.idx  \
0    999         0  nonexistent           1.1          93.994          -36.4   
1    999         0  nonexistent           1.1          93.994          -36.4   
2    999         0  nonexistent           1.1          93.994          -36.4   
3    999         0  nonexistent           1.1          93.994          -36.4   
4    999         0  nonexistent           1.1          93.994          -36.4   

   euribor3m  nr.employed   y  
0      4.857       5191.0  no  
1      4.857       5191.0  no  
2      4.857       5191.0  no  
3      4.857       5191.0  no  
4      4.857       5191.0  no

In [11]:
# We will now check if all the default = yes values do have the same result, which we believe will be "no".

# We create a boolean mask for rows where 'default' is equal to 'Yes'
mask = df['default'].isin(['yes'])

# We use the mask to filter the DataFrame and select only 'default' and 'y' columns
filtered_df = df.loc[mask, ['default', 'y']]

print(filtered_df)

      default   y
21580     yes  no
21581     yes  no
24866     yes  no


In [12]:
# We will now check if the unknown value have all the same result.
# We create a boolean mask for rows where 'default' is 'unknown'
mask = df['default'].isin(['unknown'])

# We use the mask to filter the DataFrame and select only 'default' and 'y' columns
filtered_df = df.loc[mask, ['default', 'y']]

print(filtered_df.value_counts())

default  y  
unknown  no     8154
         yes     443
Name: count, dtype: int64


It’s possible that the low number of entries with default = yes is due to a manual pre-filtering process performed by bank employees, since we only have data for those contacted during the last campaign. We could assume this is the case, which would explain the small number of 'Yes' values. However, this raises concerns about potential bias, especially due to the limited data, which could lead to data leakage during the training phase.

To address this, we will transform the default column into a binary feature, where 'Yes' and 'No' are combined into (1) and 'Unknown' becomes (0), indicating whether the default status is known. We will then assess the relevance of this variable for the model.

In [13]:
# We will specify the values in the mapping (no or yes = 1, unknown = 0).
mapping = {'no': 1, 'yes': 1, 'unknown': 0}

df['default'] = df['default'].map(mapping)

In [14]:
df['default'].value_counts()

default
1    32591
0     8597
Name: count, dtype: int64